compiler/src/test/java/com/squareup/anvil/compiler/ClassReferenceTest.kt

```diff
@@ -0,0 +1,291 @@
+package com.squareup.anvil.compiler
+
+import com.google.common.truth.Truth.assertThat
+import com.squareup.anvil.compiler.internal.fqName
+import com.squareup.anvil.compiler.internal.reference.ClassReference
+import com.squareup.anvil.compiler.internal.reference.Visibility
+import com.squareup.anvil.compiler.internal.reference.toClassReference
+import com.squareup.anvil.compiler.internal.testing.simpleCodeGenerator
+import com.tschuchort.compiletesting.KotlinCompilation.ExitCode.OK
+import org.jetbrains.kotlin.name.FqName
+import org.jetbrains.kotlin.psi.KtClassOrObject
+import org.jetbrains.kotlin.resolve.source.getPsi
+import org.junit.Test
+
+class ClassReferenceTest {
+
+  @Test fun `constructors are parsed for PSI and Descriptor APIs correctly`() {
+    compile(
+      """
+      package com.squareup.test
+
+      class SomeClass1
+
+      class SomeClass2(string: String)
+
+      class SomeClass3() {
+        constructor(string: String) : this()
+      }
+      """,
+      allWarningsAsErrors = false,
+      codeGenerators = listOf(
+        simpleCodeGenerator { psiRef ->
+          val descriptorRef = psiRef.toDescriptorReference()
+
+          when (psiRef.shortName) {
+            "SomeClass1" -> {
+              assertThat(psiRef.constructors).hasSize(0)
+
+              assertThat(descriptorRef.constructors).hasSize(1)
```

Comment:
Note that there are always differences between the PSI and Descriptor implementation, e.g. descriptors see the default constructor, but PSI doesn't.

Reply:
At one point I definitely had comments. Will add them!